src/libsyntax/ext/build.rs

```diff
@@ -844,7 +844,7 @@ impl<'a> AstBuilder for ExtCtxt<'a> {
     }
     fn pat_enum(&self, span: Span, path: ast::Path, subpats: Vec<P<ast::Pat>>) -> P<ast::Pat> {
         let pat = if subpats.is_empty() {
-            PatKind::Path(None, path)
+            PatKind::Struct(path, Vec::new(), false)
```

Comment:
>Why does pat_enum mean PatKind::TupleStruct? Can't enum variants be structs too?

The name `pat_enum` is pretty old, from times when braced enum variants `Struct{..}` were unstable and empty struct variants `Struct{}` didn't exist, it means either `TupleStruct(..)` or `Path`.

>Also for the empty case why pick empty PatKind::Struct over empty PatKind::TupleStruct?

Because it matches anything (see https://github.com/rust-lang/rfcs/pull/1506 for more details).

The alternative I mentioned above (`pat_enum` -> `pat_path`/`pat_tuple_struct`) is much cleaner at  cost of larger breakage. I will probably implement it instead.